[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/MrStrange545/ML-Engineer/blob/main/Notebooks/01_Python.ipynb)

# Table Of contents

- [1.Language Fundamentals](#1.Language-Fundamentals)
  - [1.1.Introduction](#1.1.Introduction)
  - [1.2.Identifiers](#1.2.Indentifiers)
  - [1.3.Keywords](#1.3.Keywords)
  - [1.4.Data types](#1.4.Data-Types)
     - [1.4.1.Int Data Type](#1.4.1.Int-data-type)
     - [1.4.2.Base Conversion](#1.4.2.Base-Conversion)
     - [1.4.3.Float Data Type](#1.4.3.Float-data-type)
     - [1.4.4.Complex Data Type](#1.4.4.Complex-data-type)
     - [1.4.5.Bool Data Type](#1.4.5.Bool-data-type)
     - [1.4.6.Str Data type](#1.4.6.Str-data-type)
         - [1.4.6.1.Triple Quotes String](#1.4.6.1.Triple-quotes-string)
         - [1.4.6.2.Raw String](#1.4.6.2.Raw-string)
         - [1.4.6.3.Slicing of strings](#1.4.6.3.Slicing-of-strings)
     - [1.4.7.Type Casting](#1.4.7.Type-Casting)
         - [1.4.7.1.int()](#1.4.7.1.int())
         - [1.4.7.2.float()](#1.4.
- [2.Operators](#2.Operators)









**Resources:**
1. Durga software Solutions

# 1.Language-Fundamentals

## 1.1.Introduction
- Python is a general purpose high level programming language.
- Python was developed by Guido Van Rossum in 1989 while working at National Research at netherlands.
- But officially Python was made available to public in 1991. The Official DOB: 20th feb 1991.
- The name "Python" was selected from the TV show "The Complete Monty Python's Circus", which was broadcasted in BBC from 1969 to 1974.

Guido Developed Python languages by taking almost all programming features from different languages.
- Functional Programming Features from C.
- Object Oriented Programming features from C++.
- Scripting Language Features from Perl and Shell Script.
- Modular Programming Features from Modula-3

Features of Python
- Simple and easy to learn
- Freeware and Open Source
- HLL
- Platform Independent
- Portability
- Dynamically typed
- Both OOP and POP
- Interpreted
- Extensible
- Embedded
- Extensive Library.

In [2]:
# Python program to print hello world
print("hello world")

hello world


## 1.2.Identifiers
- A name in Python program is called Identifier
- It can be class name or function name or module name or variable name.

Rules to define Identifiers in Python:
1. The only allowed characters in Python are alphabets(both case), (0 to 9) and underscore(_)
2. Identifier should not start with digit.
3. Identifiers are case sensitive.
4. we cannot use reserved words as identifiers.

**Note**:
- If identifier is start with underscore(_) then it indicates it is private
- Dollar ($) is not allowed in Python.
- There is no length limit for python identifiers, but not recommended to use too lengthy identifiers.
- If identifiers starts with __(two underscore symbols) then it indicates it is strongly private
- If the identifier starts and ends with two underscore symbols then it is a magic method.

## 1.3.Keywords
- In python some words are reserved to represent some meaning or functionality. Such type of words are called Reserved Words.
- There are 35 reserved words

**Note**: 
- All reserved words in python contain only alphabet symbols
- Except the following three all are in lower case
   - True
   - False
   - None

In [7]:
# code to print all reserved words
import keyword
keyword.kwlist

['False',
 'None',
 'True',
 'and',
 'as',
 'assert',
 'async',
 'await',
 'break',
 'class',
 'continue',
 'def',
 'del',
 'elif',
 'else',
 'except',
 'finally',
 'for',
 'from',
 'global',
 'if',
 'import',
 'in',
 'is',
 'lambda',
 'nonlocal',
 'not',
 'or',
 'pass',
 'raise',
 'return',
 'try',
 'while',
 'with',
 'yield']

## 1.4.Data-Types
- Data type represent the type of data present inside a variable.
- In python, we are not required to specify the type explicitly. Based on value provided, the type will be assigned automatically. Hence Python is `Dyanically Typed language`.

Python contains several inbuilt functions

In [8]:
# type() -> to check the type of variable or object
a = 10
type(a)

int

In [9]:
# id() -> to get address of the object
# it can change on runtime
id(a)

95101982173864

In [10]:
# print() -> to print the value of the object
print(a)

10


### 1.4.1.Int-data-type
- we can use int data type to represent integers

In [11]:
a = 9568
type(a)

int

We can represent int values in following ways
1. Decimal
2. Binary
3. Octal
4. Hexadecimal

In [15]:
# Decimal
# allowed digits are  0 to 9
# no prefix required
a = 10
a

10

In [16]:
# Binary
# allowed digits are 0 and 1
# prefixed with 0b or 0B
a = 0b111
a

7

In [17]:
# Octal
# allowed digits are 0 to 7
# prefixed with 0o or 0O
a = 0o567
a

375

In [19]:
# Hexa-Decimal
# allowed symbols are 0 to 9, and A to F
# prefixed with 0x or OX
a = 0xFACE
a

64206

Being a Programmer we can specify literal values in decimal, binary, octal, and hexa-decimal forms. But PVM will always provide values in decimal form

#### 1.4.2.Base-Conversion
- Python provide the following in-built functions for base conversions

In [20]:
# bin()
bin(15), bin(0o11), bin(0xA4), bin(0b100)

('0b1111', '0b1001', '0b10100100', '0b100')

In [21]:
# oct()
oct(10), oct(0B1111), oct(0X123)

('0o12', '0o17', '0o443')

In [22]:
# hex()
hex(100), hex(0B1111111), hex(0o12345)

('0x64', '0x7f', '0x14e5')

### 1.4.3.Float-data-type
- Used to represent decimal values

In [24]:
f = 1.234
type(f)

float

In [25]:
f = 1.2e3  # or 1.2E3
f

1200.0

We can represent int values in decimal, binary, octal and hexa decimal forms. But we can represent float values only by using decimal form.

### 1.4.4.Complex-data-type
- used to represent complex numbers

In [28]:
a = 3+5j
type(a), a

(complex, (3+5j))

In the real part if we use int value then we can specify that either by decimal,octal,binary or hexa decimal form.
But imaginary part should be specified only by using decimal form

We can even perform operations on complex data type

In [29]:
a = 10+5j
b = 2.5-2.3j
a+b, a-b, a*b, a/b

((12.5+2.7j),
 (7.5+7.3j),
 (36.5-10.5j),
 (1.1698440207972272+3.076256499133449j))

Complex data type has some inbuilt attributes to retrieve the real part and imaginary part

In [32]:
c = 10.5+3.6j
c.real, c.imag

(10.5, 3.6)

### 1.4.5.Bool-data-type
- we can use this data type to represent boolean values.
- The only allowed values are `True` and `False`.

> Internally, Python represents True as 1 and False as 0

In [35]:
b = True
type(b), b

(bool, True)

In [36]:
a, b = 10, 20
a>b

False

In [37]:
True+True, True+False

(2, 1)

### 1.4.6.Str-data-type
- str represents string data type
- A string is a sequence of characters enclosed within single quotes or double quotes.

In [40]:
s1 = 'first string'
s2 = "second string"
s1, s2

('first string', 'second string')

#### 1.4.6.1.Triple-quotes-string
Triple quotes are useful for majorly two reasons
- one for representing data in multiline
- another when we have single and double quotes in btw and we do not want to use raw string or escape characters

In [46]:
s3 = """This is a 
multiline 
string"""

s4 = """ string" that 'contains' 
many quotes"""

s3, s4

('This is a \nmultiline \nstring', ' string" that \'contains\' \nmany quotes')

#### 1.4.6.2.Raw-string
when we want our string as it is we use raw string
- to use raw string use r before the string

In [51]:
s1 = r"hello /n world"
s1

'hello /n world'

#### 1.4.6.3.Slicing-of-strings
- slice means a piece
- [] operator is called slice operator, which can be used to retrieve parts of string.
- Python strings follows zero based indexing.
- +ve index means left to right
- -ve index means right to left

In [52]:
s = 'string'

In [56]:
s[0], s[-1], s[1]

('s', 'g', 't')

In [58]:
#  s[40] this will throw index out of range error

when using `:` operator left part is starting and right part is ending(exclusive) and when using `[::]` left most part is starting index, and middle part is ending(exclusive) and right most part is step
> when nothing is described it means starting=0, end = len(string), step = 1

In [59]:
s[1:40], s[1:], s[:4], s[:]

('tring', 'tring', 'stri', 'string')

### 1.4.7.Type-Casting
- we can convert one type of value to another type. This conversion is called Typecasting or type coersion.
- The following are various inbuilt functions for type casting

#### 1.4.7.1.int()
- we can use this to convert other types to int type

In [61]:
int(123.987), int(True), int(False), int("10")

(123, 1, 0, 10)

In [62]:
# int(10+5j) typeError
# int("10.5") ValueError
# int("ten") ValueError
# int("0b1111") # valueError

**Note**:
1. We can convert any type to int except complex type.
2. If we want to convert str type to int, compulsory str should contain integral value (in decimal system)

#### 1.4.7.2.float()
- we can use `float()` to convert other types to float type

In [63]:
float(10), float(True), float(False), float("10.5")

(10.0, 1.0, 0.0, 10.5)

In [64]:
# float(10+5j) TypeError
# float("ten") valueError
# float("0b1111") valueError

**Note**:
1. We can convert any type of value to float type except complex type
2. Whenever we are trying to convert str type to float type compulsory str should be either integral or floating point literal and should be specified only in base-10

#### 1.4.7.3.complex()
we can use this `complex()` function to convert other types to complex type.

##### Form-1: complex(x)
We can use this function to convert x into complex number with real part x and imaginary part 0.

In [66]:
complex(10), complex(10.5), complex("10"), complex(True), complex("10.5")

((10+0j), (10.5+0j), (10+0j), (1+0j), (10.5+0j))

In [68]:
# complex("ten") valueError

##### Form-2: complex(x,y)
We can use this method to convert x and y into complex number such that x will be real part and y will be imaginary part.

In [69]:
complex(10,-2), complex(True, False)

((10-2j), (1+0j))

#### 1.4.7.4.bool()
- We can use this method to convert other type to bool type:

In [70]:
bool(0), bool(1), bool(10), bool(10.5), bool(0.178), bool(0.0), bool(10-2j)

(False, True, True, True, True, False, True)

In [71]:
bool("True"), bool("False"), bool("")

(True, True, False)

conversion rule for bool(x)
- If x is int:
   - 0 means False
   - non-zero means True
- If x is a float
   - If total numerical value is 0 then result is False otherwise the result is True.
- If x is a complex
   - if both real and imag are 0 then False else True
- If x is a str
   - if x is empty the result is False else True

#### 1.4.7.5.str()
- we can use this method to convert other type values to str type

In [73]:
str(10), str(10.5), str(10+5j), str(True)

('10', '10.5', '(10+5j)', 'True')

### 1.4.8.bytes-data-type
- bytes represent a group of byte numbers just like an array

In [79]:
x = [10,255,30,40]
b = bytes(x)
type(b)

bytes

In [80]:
for i in b: print(i)

10
255
30
40


**Conclusion 1:**
The only allowed values for byte data type are 0 to 256(excluded). If we are trying to provide any other values then we will get valueError.

**Conclusion 2:**
Once we creates bytes data type value, we cannot change its values, otherwise we will get TypeError.

In [82]:
# b[0] = 100 TypeError: 'bytes' object does not support item assignment

### 1.4.9.bytearray-data-type
- same as bytes except its elements can be modified

In [83]:
b = bytearray(x)
type(b)

bytearray

In [84]:
for i in b: print(i)

10
255
30
40


In [86]:
b[0] = 50

In [87]:
for i in b: print(i)

50
255
30
40


There are several advanced data structures also, we will discuss them later on in detail like list, tuple, set, dict.

### 1.4.10.range-data-type
range data type represents a sequence of numbers.
The elements present in range data type are not modifiable i.e. data type is immutable

In [89]:
# form-1: range(10)
## generate numbers from 0 to 9
r = range(10)
for i in r: print(i)

0
1
2
3
4
5
6
7
8
9


In [90]:
# form-2: range(10,20)
## generate numbers from 10 to 19
r = range(10,20)
for i in r: print(i)

10
11
12
13
14
15
16
17
18
19


In [91]:
# form-3: range(0,10,2)
## generate number from 0 to 9 with differenc of 2
r = range(0,10,2)
for i in r: print(i)

0
2
4
6
8


we can access elements present in range data type using index

In [92]:
r[2]

4

### 1.4.11.frozenset-data-type
It is same as set, but immutable
hence we cannot use add or remove functions

In [94]:
s = {10,20,30,40}
fs = frozenset(s)
type(fs), fs

(frozenset, frozenset({10, 20, 30, 40}))

### 1.4.12.None-Data-type
None means nothing or no value associated

In [95]:
def m1():
    a=10

print(m1())

None


### 1.4.13.Escape-Characters
- In string literals we can use escape characters to associate a special meaning.

| Escape Sequence | Description                          | Example       |
|-----------------|--------------------------------------|---------------|
| `\\`            | Backslash                             | `\\`          |
| `\'`            | Single quote                          | `\'`          |
| `\"`            | Double quote                          | `\"`          |
| `\n`            | Newline                               | `\n`          |
| `\r`            | Carriage return                       | `\r`          |
| `\t`            | Horizontal tab                        | `\t`          |
| `\b`            | Backspace                             | `\b`          |
| `\f`            | Form feed                             | `\f`          |
| `\v`            | Vertical tab                          | `\v`          |
| `\a`            | Bell (alert)                          | `\a`          |
| `\ooo`          | Octal value (e.g., `\007` for bell)   | `\007`        |
| `\xhh`          | Hexadecimal value (e.g., `\x41` for 'A') | `\x41`      |
| `\uhhhh`        | Unicode character (e.g., `\u03A9` for Ω) | `\u03A9`    |
| `\Uhhhhhhhh`    | Extended Unicode character (e.g., `\U0001F600` for 😀) | `\U0001F600` |


# 2.Operators

## 2.1.Arithmetic-operators

| Operator | Description          | Example                  | Result  |
|----------|----------------------|--------------------------|---------|
| `+`      | Addition             | `5 + 3`                  | `8`     |
| `-`      | Subtraction          | `10 - 4`                 | `6`     |
| `*`      | Multiplication       | `7 * 6`                  | `42`    |
| `/`      | Division             | `15 / 3`                 | `5.0`   |
| `//`     | Floor Division       | `15 // 4`                | `3`     |
| `%`      | Modulus (Remainder)  | `17 % 5`                 | `2`     |
| `**`     | Exponentiation       | `2 ** 3`                 | `8`     |

**Note:**
- / operator always performs floating point arithmetic. Hence it will always returns float value.
- // (floor division) can perform both floating point and integral arithmetic. If arguments are int type then result is int type. If atleast one argument is of float type then result is float type.

In [96]:
a, b = 10, 2
a+b, a-b, a*b, a/b, a//b, a**b, a%b

(12, 8, 20, 5.0, 5, 100, 0)

we can also use +, * for string type also.
- for + both arg must be string
- for * one arg must be int

In [98]:
'string'+'string2', 'string'*3

('stringstring2', 'stringstringstring')

## 2.2.Relational-Operators
| Operator | Description           | Example       | Result  |
|----------|-----------------------|---------------|---------|
| `==`     | Equal to              | `5 == 5`      | `True`  |
| `!=`     | Not equal to          | `5 != 3`      | `True`  |
| `>`      | Greater than          | `7 > 3`       | `True`  |
| `<`      | Less than             | `4 < 9`       | `True`  |
| `>=`     | Greater than or equal to | `5 >= 5`   | `True`  |
| `<=`     | Less than or equal to | `3 <= 4`      | `True`  |


In [100]:
a, b = 10, 20
a>b, a<b, a==b, a!=b, a>=b, a<=b

(False, True, False, True, False, True)

We can apply relational operators for str types also

In [101]:
a, b = 'str', 'str2'
a>b, a<b, a==b, a!=b, a>=b, a<=b

(False, True, False, True, False, True)

Chaining of relational operators also possible. In the chain if result is False at any step then the whole result will become False else True

In [102]:
10>20, 10<20<30, 10<20<30<40, 10<20<30>40<50

(False, True, True, False)

## 2.3.Logical-Operators
- we can apply logical operators for all types

This is for Boolean type
| Operator | Description         | Example               | Result |
|----------|---------------------|-----------------------|--------|
| `and`    | Logical AND         | `True and False`      | `False`|
| `or`     | Logical OR          | `True or False`       | `True` |
| `not`    | Logical NOT         | `not True`            | `False`|


This is for Non-Boolean type
| Operator | Description         | Example                | Result        |
|----------|---------------------|------------------------|---------------|
| `and`    | Logical AND         | `0 and 5`              | `0`           |
| `or`     | Logical OR          | `0 or 5`               | `5`           |
| `not`    | Logical NOT         | `not 0`                | `True`        |

**Rules:**
- x and y
  - if x is false result is x else result is y
- x or y
  - if x is false result is y else result is x
-  not x
   - if x is false result is true else true

## 2.4.Bitwise-operators
- these are applicable only on boolean and int values.

For boolean same as-> logical operators

For integers, we have to convert every integer into bits then do the operations for each bit and its corresponding bit

| Operator | Description          | Example     | Result |
|----------|----------------------|-------------|--------|
| `&`      | Bitwise AND          | `5 & 3`     | `1`    |
| `|`    | Bitwise OR           | `5 | 3`     | `7`    |
| `^`      | Bitwise XOR          | `5 ^ 3`     | `6`    |
| `~`      | Bitwise NOT          | `~5`        | `-6`   |
| `<<`     | Left Shift           | `5 << 1`    | `10`   |
| `>>`     | Right Shift          | `5 >> 1`    | `2`    |

bitwise complement uses 2`s complement.
```
print(~5) ==> -6
```

## 2.5.Assignment-Operators
| Operator | Description                   | Example       | Result   |
|----------|-------------------------------|---------------|----------|
| `+=`     | Addition assignment           | `a += 3`      | `a = a + 3` |
| `-=`     | Subtraction assignment        | `a -= 2`      | `a = a - 2` |
| `*=`     | Multiplication assignment     | `a *= 4`      | `a = a * 4` |
| `/=`     | Division assignment           | `a /= 2`      | `a = a / 2` |
| `//=`    | Floor division assignment     | `a //= 3`     | `a = a // 3`|
| `%=`     | Modulus assignment            | `a %= 5`      | `a = a % 5` |
| `**=`    | Exponentiation assignment     | `a **= 2`     | `a = a ** 2`|
| `&=`     | Bitwise AND assignment         | `a &= 3`      | `a = a & 3` |
| `|=`     | Bitwise OR assignment          | `a |= 2`      | `a = a | 2` |
| `^=`     | Bitwise XOR assignment         | `a ^= 1`      | `a = a ^ 1` |
| `<<=`    | Left shift assignment          | `a <<= 1`     | `a = a << 1`|
| `>>=`    | Right shift assignment         | `a >>= 2`     | `a = a >> 2`|


## 2.6.Ternary-operator
syntax:
```
x = first value if condition else second value
```
Nesting of ternary operator is possible

In [107]:
a,b = 10, 20
result = a if a>b else b
result

20

## 2.7.Special-operators
#### 2.7.1.Identity-Operators
1. is
2. is not

- r1 is r2 returns True if both r1 and r2 are pointing to the same object.
- r1 is not r2 returns True if both r1 and r2 are not pointing to the same objec

In [108]:
a,b = 10,10
a is b, a is not b

(True, False)

In [110]:
list1 = [1,2,3]
list2 = [1,2,3]
list1 is list2, list1 == list2

(False, True)

we can use is operator for address comparision where as == operator for content comparision.

### 2.7.2.Membership-operators:
1. in 
2. not in

- in -> Returns True if the given object present in the specified Collection
- not in -> Retruns True if the given object not present in the specified Collection

In [111]:
list1 = [1,2,3]
1 in list1, 5 not in list1

(True, True)

## 2.8.Operator-Precedence

| Precedence | Operator(s)                | Description                                   | Example               |
|------------|----------------------------|-----------------------------------------------|-----------------------|
| 1          | `()`                       | Parentheses                                   | `(2 + 3) * 4`         |
| 2          | `**`                       | Exponentiation                                | `2 ** 3`              |
| 3          | `+x, -x, ~x`               | Unary plus, Unary minus, Bitwise NOT          | `-3 ** 2`, `+x`, `~x` |
| 4          | `*`, `/`, `//`, `%`        | Multiplication, Division, Floor Division, Modulus | `5 * 2`, `10 / 3`, `10 // 3`, `10 % 3` |
| 5          | `+`, `-`                   | Addition, Subtraction                         | `2 + 3`, `5 - 2`      |
| 6          | `<<`, `>>`                 | Bitwise Shift Left, Bitwise Shift Right       | `2 << 1`, `8 >> 2`    |
| 7          | `&`                        | Bitwise AND                                   | `5 & 3`               |
| 8          | `^`                        | Bitwise XOR                                   | `5 ^ 3`               |
| 9          | `|`                        | Bitwise OR                                    | `5 | 3`               |
| 10         | `==`, `!=`, `>`, `<`, `>=`, `<=` | Equality, Inequality, Greater Than, Less Than, Greater Than or Equal To, Less Than or Equal To | `5 == 5`, `3 < 4`     |
| 11         | `and`                      | Logical AND                                   | `True and False`      |
| 12         | `or`                       | Logical OR                                    | `True or False`       |
| 13         | `not`                      | Logical NOT                                   | `not True`            |
| 14         | `=`                        | Assignment                                    | `x = 5`               |
| 15         | `+=`, `-=`, `*=`, `/=`, `//=`, `%=` | Augmented Assignment                    | `x += 2`              |
| 16         | `,`                        | Comma (used in tuples, lists, and function arguments) | `(1, 2, 3)`, `func(a, b)` |

# 3.Math-Module
- A module is a collection of functions, variables and classes, etc.
- Math is a module that contains several functions to perform mathematical operations
- if we want to use that module then we have to import it first.

In [113]:
import math as m
m.sqrt(16), m.pi

(4.0, 3.141592653589793)

Common functions
| Function        | Description                                                                                 |
|-----------------|---------------------------------------------------------------------------------------------|
| `math.abs(x)`   | Returns the absolute value of `x`.                                                          |
| `math.ceil(x)`  | Returns the smallest integer greater than or equal to `x`.                                  |
| `math.floor(x)` | Returns the largest integer less than or equal to `x`.                                      |
| `math.factorial(x)` | Returns the factorial of `x`, where `x` is a non-negative integer.                        |
| `math.fabs(x)`  | Returns the absolute value of `x` as a float.                                               |
| `math.sqrt(x)`  | Returns the square root of `x`.                                                              |
| `math.pow(x, y)`| Returns `x` raised to the power of `y`.                                                      |
| `math.exp(x)`   | Returns `e` raised to the power of `x`.                                                      |
| `math.log(x, base)` | Returns the logarithm of `x` to the specified `base`. If `base` is not provided, returns the natural logarithm. |
| `math.log10(x)` | Returns the base-10 logarithm of `x`.                                                        |
| `math.sin(x)`   | Returns the sine of `x`, where `x` is in radians.                                            |
| `math.cos(x)`   | Returns the cosine of `x`, where `x` is in radians.                                          |
| `math.tan(x)`   | Returns the tangent of `x`, where `x` is in radians.                                        |
| `math.degrees(x)` | Converts angle `x` from radians to degrees.                                                |
| `math.radians(x)` | Converts angle `x` from degrees to radians.                                                |
| `math.trig(x)`  | Returns the trigonometric function of `x` (e.g., `sin`, `cos`, `tan`).                      |
| `math.isqrt(x)` | Returns the integer square root of `x`, where `x` is a non-negative integer.                |
| `math.gcd(x, y)`| Returns the greatest common divisor of `x` and `y`.                                         |
| `math.hypot(x, y)` | Returns the Euclidean norm, `sqrt(x*x + y*y)`.                                               |
| `math.atan2(y, x)` | Returns the arc tangent of `y / x` in radians, handling all quadrants.                     |
| `math.asin(x)`  | Returns the arc sine of `x` in radians.                                                      |
| `math.acos(x)`  | Returns the arc cosine of `x` in radians.                                                    |
| `math.sinh(x)`  | Returns the hyperbolic sine of `x`.                                                           |
| `math.cosh(x)`  | Returns the hyperbolic cosine of `x`.                                                         |
| `math.tanh(x)`  | Returns the hyperbolic tangent of `x`.                                                        |


Common constants
| Constant       | Description                                               | Value                |
|----------------|-----------------------------------------------------------|----------------------|
| `math.pi`      | The mathematical constant π (pi), approximately 3.14159.  | `3.141592653589793`  |
| `math.e`       | The mathematical constant e (Euler's number), approximately 2.71828. | `2.718281828459045`  |
| `math.tau`     | The mathematical constant τ (tau), which is 2π, approximately 6.28318. | `6.283185307179586`  |
| `math.inf`     | Represents positive infinity.                             | `float('inf')`       |
| `math.nan`     | Represents Not-a-Number (NaN).                            | `float('nan')`       |
| `math.nan`     | Represents a Not-a-Number (NaN) value.                    | `float('nan')`       |


# 4.Flow-Control


## 4.1.Conditional-statements
- everything is same as java.

**Note**:
- else part is always optional
- hence the various possible syntaxes are:
  - if
  - if - else
  - if - elif - else
  - if - elif

## 4.2.Iterative-statements
1. for loop
2. while loop

### 4.2.1.for-loop
syntax:
for x in sequence:
    body

where sequence can be string or any collection.
body will be executed for every element present in the sequence.

while loop is exactly same as java.

## 4.3.Transfer-Statements
### 4.3.1.break
same as java

### 4.3.2.continue
same as java

### 4.3.3.loops-with-else-block
Inside loop execution, if break statement is not executed then only else part will be executed.
> else means loop without break

**Questions:**
1. What is the difference btw for and while loop?
we can use loops to repeat code execution
repeat code for every item in sequence --> for loop
repeat code as long as condition is true --> while loop

2. How to exit from the loop?
by using break statement

3. How to skip some iteration inside loop?
by using continue statement

4. when else part will be executed wrt loops?
if loop executed without break


### 4.3.4.pass-statement
- pass is a keyword in python
- In our programming syntactically if block is required which won't do anything then we can define that empty block with pass keyword


In [116]:
if True:
    pass

In [117]:
def m1():
    pass

### 4.3.5.del-statement
After using a variable, it is highly recommended to delete that variable if it is no longer required,so that the corresponding object is eligible for Garbage Collection. We can delete variable by using del keyword

In [118]:
x = 10
del x

# 5.String
- any sequence of characters within either single quotes or double quotes.

**How to define multi-line string literals:**
using triple quotes

**How to access characters of a string:**
1. by using index
2. by using slice operator

**Built-in Functions**
1. len():
We can use len() to find the number of characters present in the string.

**Checking Membership:**

In [119]:
s = 'string'
's' in s, 't' in s, 'd' in s

(True, True, False)

**Comparsion of strings**
- we can use comparision operators ( <, <=, >, >=, ==, != ) for strings
- comparision will be performed based on alphabetical order.

## 5.1.Removing-spaces
We can use the following 3 methods
1. rstrip() -> To remove spaces at right hand side
2. lstrip() -> To remove spaces at left hand side
3. strip()  -> To remove spaces both sides

## 5.2.Finding-substrings:
We can use the following 4 methods 
### for forward direction:
find()
index()

### for backward direction:
rfind()
index()

### 1. find()
```
s.find(substring)
```
Returns index of first occurence of the given substring. If it is not available then we will get -1.

In [122]:
s='Learning Python is very easy'
s.find("Python"), s.find("Java"), s.find('r'), s.rfind('r')

(9, -1, 3, 21)

**Note:** By default find() method can search total string. We can also specify the boundaries to search.
```
s.find(substring, begin_index, end_index)
```
It will always search from begin index to end-1 index

In [123]:
s='durgaravipavanshiva'
s.find('a'), s.find('a',7,15), s.find('z',7,15)

(4, 10, -1)

## 5.3.index() method
index() method is exactly same as find() method except that if the specified substring is not available then we will get valueError.

In [127]:
s = "Learning python is very easy"
subs = 'python'
s.index(subs)

9

## 5.4.Count
We can find the number of occurences of substring present in the given using count() method.
```
s.count(substring) -> it will search through out the string
s.count(substring, begin_index, end_index) -> it will search from begin index to end-1 index
```

In [128]:
s = "abcavcavcavcavccc"
s.count('a'), s.count('av'), s.count('a', 3,7)

(5, 4, 2)

## 5.5.replace
```
s.replace(oldstring, newstring)
```
inside s, every occurrence of oldstring will be replaced with newstring.

In [133]:
s = "Learning Python is very difficult"
s.replace("difficult", "easy")

'Learning Python is very easy'

#### Q. String object are immutable then how can we change the content by using replace() method.
Once we creates string object, we cannot change the content. This non changeable behaviour is nothing but immutability. If we are trying to change the content by using any method, then with those changes a new object will be created and changes won't be happened in exisiting object.

Hence with replace() method also a new object got created but existing object won't changed.

## 5.6.split()
we can split the string according to specified separator by using split() method.
```
l = s.split(separator)
```
The default separator is space. The return type of split() method is list.

In [134]:
s = "durga software solutions"
l = s.split()
for x in l: print(x)

durga
software
solutions


## 5.7.join()
we can join a group of strings(list or tuple) wrt the given separator.
```
s = separator.join(group of strings)
```

In [135]:
list1 = ['1','2']
s = ':'.join(list1)
s

'1:2'

## 5.8.Changing-case
We can change case of a string by using the following methods.
1. `upper()`-> To convert all characters into upper case
2. `lower()`-> to convert all characters into lower case
3. `swapcase()`-> to change the case
4. `title()`-> first character of every word is captial rest are lower case.
5. `capitalize()`-> only first character of string is in captial rest are in lower case.

In [137]:
s = "learning Python is very Easy"
s.upper(), s.lower(), s.swapcase(), s.title(), s.capitalize()

('LEARNING PYTHON IS VERY EASY',
 'learning python is very easy',
 'LEARNING pYTHON IS VERY eASY',
 'Learning Python Is Very Easy',
 'Learning python is very easy')

## 5.9.Checking-starting-and-ending
```
s.startswith(substring)
s.endswith(substring)
```

In [138]:
s = "learning python is very easy"
s.startswith('learning'), s.endswith('easy')

(True, True)

## 5.10 Checking-type-of-characters:
1. `isalnum()` Returns true if all characters are alphanumeric
2. `isalpha()` Returns true if all characters are only alphabets.
3. `isdigit()` Returns true if all characters are digits only.
4. `islower()` Returns true if all characters are in lower case.
5. `isupper()` Returns true if all characters are in upper case.
6. `istitle()` Returns true if string is in title case.
7. `isspace()` Returns true if string contains only spaces.

## 5.11.Formatting-the-strings
### 5.11.1.Formatting-variables

In [139]:
name = 'durga'
salary = 10000
age = 48
print(f"{name}'s salary is {salary} and his age is {age}")

durga's salary is 10000 and his age is 48


### 5.11.2.Formatting-numbers
- d -> decimal integer
- f -> fixed floating point. The default is 6.
- b -> binary format
- o -> octal format
- x -> hexadecimal format (lower-case)
- X -> hexadecimal format (upper-case)

In [3]:
# printing Integer values
print(f"The integer number is {123}") # simple integer
print(f"The integer number is {123:d}") # integer with default digits
print(f"The integer number is {123:5d}") # int with 5 digits
print(f"The integer number is {123:05d}") # int with 5 digits (with 0)

The integer number is 123
The integer number is 123
The integer number is   123
The integer number is 00123


In [4]:
# printing decimal values
print(f"The float number is {12.34}") # simple floating number
print(f"The float number is {12.34:f}") # default precision (6)
print(f"The float number is {12.34:8.3f}") # 8 digits with 3 after point
print(f"The float number is {12.34:08.3f}") # same as above but with 0
print(f"The float number is {123.4567:08.3f}") # round off 
print(f"The float number is {786786123.45:08.3f}") 

The float number is 12.34
The float number is 12.340000
The float number is   12.340
The float number is 0012.340
The float number is 0123.457
The float number is 786786123.450


**Note**
{:08.3f}

- total positions should be minimum 8
- After the decimal point exactly 3 digits are allowed. If it is less then 0s will be places in the last positions.
- If total number is <8 positions then 0 will be places in MSBx
- If total number is >8 positions then all integral digits will be considered.
- The extra digit we can take is only 0.

In [7]:
# printing decimal(base-10) values in other number systems
print(f"Decimal Form: {153}")
print(f"Binary Form: {153:b}")
print(f"Octal Form: {153:o}")
print(f"Hexadecimal Form: {154:x}")
print(f"Hexadecimal Form: {154:X}")

Decimal Form: 153
Binary Form: 10011001
Octal Form: 231
Hexadecimal Form: 9a
Hexadecimal Form: 9A


### 5.11.3.Formatting-Signed-integers

In [9]:
print(f"int value with +ve sign {123:+d}")
print(f"int value with -ve sign {-123:+d}")
print(f"float value with +ve sign {123.23:+f}")
print(f"float value with -ve sign {-12.34:+f}")

int value with +ve sign +123
int value with -ve sign -123
float value with +ve sign +123.230000
float value with -ve sign -12.340000


# 6.List
- If we want to represent a group of individual objects as a single entity, then use List.
Properties of List:
- Insertion order is preserved
- duplicate objects are allowed
- heterogenous objects are allowed
- List is growable in nature
- List supports both +ve and -ve index

## 6.1.Creation-of-list-objects
### 6.1.1.Empty-list

In [4]:
l = []
l2 = list()
l, l2,  type(l), type(l2)

([], [], list, list)

### 6.1.2.Static-input

In [2]:
l = [10,20,30,40]
l

[10, 20, 30, 40]

### 6.1.3.Dynamic-input

In [5]:
l = eval(input("Enter list:"))
l, type(l)

Enter list: [10,20,30]


([10, 20, 30], list)

### 6.1.4.List-function

In [6]:
l = list(range(0,10,2))
l, type(l)

([0, 2, 4, 6, 8], list)

In [7]:
s = "durga"
l = list(s)
l, type(l)

(['d', 'u', 'r', 'g', 'a'], list)

### 6.1.5.With-split-function

In [8]:
s = "Learning Python is very Easy !!!"
l = s.split()
l, type(l)

(['Learning', 'Python', 'is', 'very', 'Easy', '!!!'], list)

**NOTE:**
>Sometimes we can take list inside another list, such type of lists are called nested lists.

## 6.2.Accesing-elements-of-list
We can access elements of the list either by using index or using slice operator(:)

### 6.2.1.By-using-index:
- List follows zero based index (i.e. index of first element is zero.)
- List supports both +ve and -ve indices
- +ve index meant for left to right
- -ve index meant for right to left

In [11]:
l = list(range(10,41,10))
l[0], l[-1]

(10, 40)

In [12]:
# l[10] IndexError: list index out of range

### 6.2.2.By-using-slice-operator:
l2=l1[start:stop:step]

```
start => it indicates the index where slice has to start.
         default value if 0

Stop => it idicates the index where slice has to end.
        default value is max, allowed index of list i.e. len of the list
        
Step => increment value.
        default is 1
```

In [32]:
l = list(range(10,100,10))
print(l)
print(l[2:7:2])
print(l[4:100])

[10, 20, 30, 40, 50, 60, 70, 80, 90]
[30, 50, 70]
[50, 60, 70, 80, 90]


## 6.3.List-vs-immutability
Once we creates a list object, we can modify its content, Hence, list objects are mutable.

## 6.4.List-methods

### 6.4.1.len()
returns the number of element present in the list

In [33]:
len(l)

9

### 6.4.2.count()
returns the number of occurence of specified item in the list.

In [34]:
l.count(20)

1

### 6.4.3.index()
returns the index of first occurance of the specified item

In [35]:
l.index(60)

5

If the item not present in list then we will get valueError

### 6.4.4.append()
we can use append() function to add item at the end of the list.

In [36]:
l = list()
l.append(10)
l.append(20)
l

[10, 20]

### 6.4.5.insert()
to insert item at specified index position

In [37]:
l.insert(1, 888) # first index, then value
l

[10, 888, 20]

**NOTE:** If the specified index is greater than max index then element will be inserted at the position. If the specified index is smaller than min index then element will be inserted at first position.

### 6.4.6.extend()
to add all items of one list to another list

In [38]:
tmp_list = list(range(1,10,2))
l.extend(tmp_list)
l

[10, 888, 20, 1, 3, 5, 7, 9]

### 6.4.7.remove()
- to remove specified item from the list
- if the item is present multiple times then only first occurance will be removed

In [39]:
l.remove(20)
l

[10, 888, 1, 3, 5, 7, 9]

**NOTE:**
If the specified item not present in the list then we will get valueError.

### 6.4.8.pop()
It removes and returns the last element of the list
This is only function which manipulates list and returns some element.

In [40]:
l.pop()
l

[10, 888, 1, 3, 5, 7]

**NOTE**: 
- If the list is empty then pop will raise error
- pop() is the only function which manipulates the list and returns some value
- In general, we can use append() and pop() functions to implement stack data structure by using list, which follows LIFO(Last In First Out) order.
- In general we can use pop() function to remove last element  of the list. But we can use to remove elements based on index.

- n.pop(index) ==> To remove and return element present at specified index.
- n.pop() ==> to remove and return last element of the list

In [41]:
l = [10,20,30,40,50,60]
l.pop(), l.pop(2)

(60, 30)

### 6.4.9.reverse()
we can use to reverse() order of elements of list

In [45]:
print(l)
l.reverse()
print("Reversed list:",l)

[10, 20, 40, 50]
Reversed list: [50, 40, 20, 10]


### 6.4.10.sort()
- In list by default insertion order is preserved. If want to sort the elements of list according to default natural sorting order then we should go for sort() method.
- For numbers ==> default natural sorting order is Ascending Order
- For strings ==> default natural sorting order is Alphabetical Order

In [51]:
print(l)
l.sort()
print("Sorted numerical list:",l)

s = ['dog','cat','banana','apple']
print(s)
s.sort()
print("sorted string list",s)

[10, 20, 40, 50]
Sorted numerical list: [10, 20, 40, 50]
['dog', 'cat', 'banana', 'apple']
sorted string list ['apple', 'banana', 'cat', 'dog']


**NOTE:**
- to use sort() function, compulsory list should contain only homogeneous elements. Otherwise we will get TypeError.

To sort in reverse from default natural order:
- use attribute `reverse=True`

In [52]:
s.sort(reverse=True)
s

['dog', 'cat', 'banana', 'apple']

### 6.4.11.clear()
we can use clear() function to remove all elements of list

In [69]:
l = [10,20,30,40]
l.clear()
l

[]

## 6.5.Aliasing-and-cloning-of-list-objects
>The process of giving another reference variable to the existing list is called aliasing.

In [54]:
x = [10,20,30,40]
y=x
id(x) == id(y)

True

The problem in this approach is by using one reference variable if we are changing content, then those changes will be reflected to the other reference variable.

In [55]:
x[0] =888
y

[888, 20, 30, 40]

To overcome this problem we should go for cloning.
>The process of creating exactly duplicate independent object is called cloning

We can implement cloning by using slice operator or by using `copy()` function.

In [56]:
x = [10,20,30,40]
y = x[:]
id(x) ==id(y)

False

In [57]:
x = [10,20,30,40]
y = x.copy()
id(x) ==id(y)

False

Q.What is the difference btw = and copy() function?
- = meant for aliasing
- copy() function meant for cloning

## 6.6.Operations-on-list
### 6.6.1.Concatenation

In [59]:
a = [10,20,30]
b = [40,50,60]
a+b

[10, 20, 30, 40, 50, 60]

**NOTE:**
- to use + operator compulsory both arguments should be list objects, otherwise we will get typeError.

In [62]:
# a+40 this will give error
# to add some integer in the list first convert it into list
a+[40]

[10, 20, 30, 40]

### 6.6.2.Repetition-operator
- we can use repetition operator * to repeat element of the list specified number of times.

In [63]:
x = [10,20,30]
y = x*3
y

[10, 20, 30, 10, 20, 30, 10, 20, 30]

## 6.7.Comparing List Objects

In [64]:
x = ['Dog', 'Cat', 'Rat']
y = ['Dog', 'Cat', 'Rat']
z = ['DOG', 'CAT', 'RAT']
x==y, x==z, x!=z

(True, False, True)

**NOTE:**
Whenever we are using comparision operators (==, !=) for list objects then the following should be considered.
1. The number of elements.
2. The order of elements.
3. The content of elements (case sensitive)

**NOTE:**
- whenever we are using relational operators (<, <=, >, >=) between list objects, only first element comparision will be performed. 

In [67]:
x = [80,20,30]
y = [70,80,90]
x>y, x>=y, x<y

(True, True, False)

## 6.8.Membership-operator
- we can check whether element is a member of the list or not by using membership operators.
    - in
    - not in

In [68]:
l = [10,50,60,30,20]
10 in l, 90 in l

(True, False)

## 6.9.List-comprehensions
- It is very easy and compact way of creating list objects from any iterable objects(like list, tuple, dict, range, etc) based on some condition.
Syntax
```
list=[expression for item in list if condition]
```

In [70]:
s = [x*x for x in range(1,11)]
s

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

In [73]:
m = [x**0.5 for x in s if x%2==1]
m

[1.0, 3.0, 5.0, 7.0, 9.0]

In [74]:
words = ["hello", "hi", "Chiranjeevi", "Nageshwar"]
l = [w[0] for w in words]
l

['h', 'h', 'C', 'N']

# 7.Tuple
Tuple is exactly same as List except that it is immutable. i.e. once we creates Tuple object, we cannot perform any changes in that object.
- tuple is read version of list
- If our data is fixed and never changes then we should go for tuple.
- Insertion order is preserved
- Duplicates are allowed
- Heterogeneous objects are allowed
- we can preserve insertion order and we can differentiate duplicate objects by using index. Hence, index will play important role in Tuple also.
- Tuple supports both +ve and -ve index. +ve index means forward direction and -ve index backward direction.
- We can represent tuple elements within paranthesis and with comma separator `parenthesis are optional` but recommended to use.

In [1]:
t = 10,20,30,40
t, type(t)

((10, 20, 30, 40), tuple)

**NOTE:**
- we have to take special care about single valued tuple.
- compulsary the value should ends with comma, otherwise it is treated as tuple.

In [2]:
t = (10,)
t, type(t)

((10,), tuple)

## 7.1.Tuple-creation
### 7.1.1.Empty-tuple

In [8]:
t = ()
t, type(t)

((), tuple)

In [9]:
t = tuple()
t, type(t)

((), tuple)

### 7.1.2.Static-input

In [5]:
t = (10,)
t, type(t)

((10,), tuple)

In [6]:
t = 10,20,30
t1 = (10,20,30)
t, t1, type(t), type(t1)

((10, 20, 30), (10, 20, 30), tuple, tuple)

### 7.1.2.By-sequences
we can create from any seq like list, range, set with the help of `tuple()` function.

In [11]:
l = [10,20,30,40]
t = tuple(l)
t, type(t)

((10, 20, 30, 40), tuple)

In [13]:
t = tuple(range(10,100,10))
t

(10, 20, 30, 40, 50, 60, 70, 80, 90)

## 7.2.Accessing-elements-of-tuple
we can access either by index or by slice operator.

### 7.2.1.By-using-index

In [16]:
t = (10,20,30,40)
t[0], t[-1]

(10, 40)

In [17]:
# t[100] IndexError: tuple index out of range

### 7.2.2.By-using-slice-operator

In [18]:
t = (10,20,30,40,50)

t[2:5], t[2:100]

((30, 40, 50), (30, 40, 50))

**Tuple vs immutability**
Once we creates tuple, we cannot change its content.
Hence tuple objects are immutable.

In [27]:
import numpy as np
a = np.array([[1,2,3],[4,5,6]])
b = np.array([[7,8],[9,10],[11,12]])
np.matmul(a,b)

array([[ 58,  64],
       [139, 154]])